gcc/rust/parse/rust-parse-impl.h

```diff
@@ -9670,6 +9670,17 @@ Parser<ManagedTokenSource>::parse_bare_function_type (
   if (!skip_token (FN_KW))
     return nullptr;
 
+  auto t = lexer.peek_token ();
+  if (t->get_id () != LEFT_PAREN)
+    {
+      Error error (t->get_locus (),
+		   "unexpected token %qs - expected bare function",
+		   t->get_token_description ());
+      add_error (std::move (error));
+
+      return nullptr;
+    }
+
   if (!skip_token (LEFT_PAREN))
     return nullptr;
```

Comment:
This condition can be removed, we've already checked for a `LEFT_PAREN` and returned a nullptr, we only need to skip it.